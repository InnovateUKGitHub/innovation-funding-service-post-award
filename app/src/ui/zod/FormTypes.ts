enum FormTypes {
  // MSP Document Share page
  ProjectLevelDelete = "projectLevelDelete",
  PartnerLevelDelete = "partnerLevelDelete",
  ProjectLevelUpload = "projectLevelUpload",

  // Claim page
  ClaimLevelDelete = "claimLevelDelete",
  ClaimLevelUpload = "claimLevelUpload",
  ClaimForecastSaveAndContinue = "claimForecastSaveAndContinue",
  ClaimForecastSaveAndQuit = "claimForecastSaveAndQuit",
  ClaimLineItemSaveAndDocuments = "claimLineItemSaveAndDocuments",
  ClaimLineItemSaveAndQuit = "claimLineItemSaveAndQuit",

  // Project Setup
  ProjectSetupBankDetails = "projectSetupBankDetails",
  ProjectSetupContactAssociate = "projectSetupContactAssociate",
  ProjectSetupForecast = "projectSetupForecast",

  // Forecast Tile
  ForecastTileForecast = "forecastTileForecast",

  // Claim Details page
  ClaimDetailLevelUpload = "claimDetailLevelUpload",
  ClaimDetailLevelDelete = "claimDetailLevelDelete",

  // Claim review page
  ClaimReviewLevelSaveAndContinue = "claimReviewSaveAndContinue",
  ClaimReviewLevelUpload = "claimReviewLevelUpload",
  ClaimReviewLevelDelete = "claimReviewLevelDelete",

  // Project Change Requests
  ProjectChangeRequestCreate = "projectChangeRequestCreate",
  ProjectChangeRequestUpdateTypes = "projectChangeRequestUpdateTypes",

  PcrLevelDelete = "pcrLevelDelete",
  PcrLevelUpload = "pcrLevelUpload",

  PcrFinancialVirementsSummary = "pcrFinancialVirementsSummary",
  PcrFinancialVirementsCostCategorySaveAndContinue = "pcrFinancialVirementsCostCategorySaveAndContinue",
  PcrReallocateCostsChangeRemainingGrant = "pcrReallocateCostsChangeRemainingGrant",

  PcrApproveNewSubcontractorStep = "pcrApproveNewSubcontractorDetails",
  PcrApproveNewSubcontractorSummary = "pcrApproveNewSubcontractorSummary",

  PcrAddPartnerCompaniesHouseStepSaveAndContinue = "pcrAddPartnerCompaniesHouseStepSaveAndContinue",
  PcrAddPartnerCompaniesHouseStepSaveAndQuit = "pcrAddPartnerCompaniesHouseStepSaveAndQuit",

  PcrChangeProjectScopeProposedPublicDescriptionStepSaveAndContinue = "pcrChangeProjectScopeProposedPublicDescriptionStepSaveAndContinue",
  PcrChangeProjectScopeProposedProjectSummaryStepSaveAndContinue = "pcrChangeProjectScopeProposedProjectSummaryStepSaveAndContinue",
  PcrChangeProjectScopeSummary = "pcrChangeProjectScopeSummary",

  PcrRenamePartnerStep = "pcrRenamePartnerStep",
  PcrRenamePartnerFilesStep = "pcrRenamePartnerFilesStep",
  PcrRenamePartnerSummary = "pcrRenamePartnerSummary",

  PcrPrepareReasoningStep = "pcrPrepareReasoningStep",
  PcrPrepareReasoningFilesStep = "pcrPrepareReasoningFilesStep",
  PcrPrepareReasoningSummary = "pcrPrepareReasoningSummary",

  LoanLevelUpload = "loanLevelUpload",
  LoanLevelDelete = "loanLevelDelete",
  LoanRequest = "loanRequest",
}

export { FormTypes };
