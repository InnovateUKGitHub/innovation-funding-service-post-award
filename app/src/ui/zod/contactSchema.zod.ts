import { z } from "zod";
import { dateValidation, pclIdValidation } from "./helperValidators/helperValidators.zod";
import { makeZodI18nMap } from "@shared/zodi18n";

const contactSchemaErrorMap = makeZodI18nMap({ keyPrefix: ["contact"] });

const contactDtoSchema = z.object({
  id: pclIdValidation,
  associateStartDate: dateValidation,
});

const multipleContactDtoSchema = z.object({
  contacts: z.array(contactDtoSchema),
});

type MultipleContactDtoSchemaType = typeof multipleContactDtoSchema;
type ContactDtoSchemaType = typeof contactDtoSchema;

export {
  contactDtoSchema,
  multipleContactDtoSchema,
  contactSchemaErrorMap,
  ContactDtoSchemaType,
  MultipleContactDtoSchemaType,
};
