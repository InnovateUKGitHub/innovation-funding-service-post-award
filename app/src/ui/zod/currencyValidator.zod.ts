import { parseCurrency, validCurrencyRegex } from "@framework/util/numberHelper";
import { isNil } from "lodash";
import { z, ZodIssueCode } from "zod";

const defaultMaxCurrency = 999_999_999_999;
const defaultMinCurrency = 0;

export const getGenericCurrencyValidation = ({
  label,
  max = defaultMaxCurrency,
  min = defaultMinCurrency,
  required = false,
}: {
  label: string;
  max?: number;
  min?: number;
  required?: boolean;
}) => {
  return z
    .string()
    .nullable()
    .superRefine((val, ctx) => {
      const currency = parseCurrency(val);
      const isEmpty = (typeof val === "string" && val.trim() === "") || isNil(val);
      if (required && isEmpty) {
        return ctx.addIssue({
          code: ZodIssueCode.custom,
          params: {
            label,
            generic: ["currency"],
            i18n: "errors.required",
          },
        });
      }

      if (!isEmpty) {
        if (/[\$€¥]+/.test(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.not_pounds",
            },
          });
        }
        // Check if the string can even be parsed
        if (isNaN(currency)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.not_a_number",
            },
          });
        }

        // Make sure our currency isn't so big as to break our server
        if (currency > max) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              count: max,
              generic: ["currency"],
              i18n: "errors.too_big",
            },
          });
        }

        if (currency < min) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              count: min,
              generic: ["currency"],
              i18n: "errors.too_small",
            },
          });
        }
        if (/\.\d\d\d/.test(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.two_decimal_places",
            },
          });
        }

        if (!validCurrencyRegex.test(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.invalid_currency",
            },
          });
        }
      }
    })
    .transform(x => {
      if ((typeof x === "string" && x.trim() === "") || isNil(x)) {
        return null;
      }
      return parseCurrency(x);
    });
};
