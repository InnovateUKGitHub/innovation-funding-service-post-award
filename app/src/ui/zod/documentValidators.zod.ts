import {
  allowedClaimDocuments,
  allowedImpactManagementClaimDocuments,
  allowedProjectLevelDocuments,
  DocumentDescription,
  allowedPcrLevelDocuments,
} from "@framework/constants/documentDescription";
import { IAppOptions } from "@framework/types/IAppOptions";
import { z } from "zod";
import { FormTypes } from "./FormTypes";
import {
  projectIdValidation,
  emptyStringToUndefinedValidation,
  partnerIdValidation,
  getMultiFileValidation,
  periodIdValidation,
  pcrItemIdValidation,
  costCategoryIdValidation,
  pcrIdValidation,
} from "./helperValidators.zod";
import { ProjectDto } from "@framework/dtos/projectDto";
import { ImpactManagementParticipation } from "@framework/constants/competitionTypes";
import { makeZodI18nMap } from "@shared/zodi18n";

type ProjectLevelUploadSchemaType = ReturnType<typeof getProjectLevelUpload>;
const getProjectLevelUpload = (config: IAppOptions) =>
  z.object({
    form: z.literal(FormTypes.ProjectLevelUpload),
    projectId: projectIdValidation,
    partnerId: z.union([emptyStringToUndefinedValidation, partnerIdValidation]),
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x => allowedProjectLevelDocuments.includes(x))
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    files: getMultiFileValidation(config),
  });

type UploadBankStatementSchemaType = ReturnType<typeof getBankStatementUpload>;
const getBankStatementUpload = (config: IAppOptions) =>
  z.object({
    form: z.literal(FormTypes.ProjectLevelUpload),
    projectId: projectIdValidation,
    partnerId: z.union([emptyStringToUndefinedValidation, partnerIdValidation]),
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x => [DocumentDescription.BankStatement].includes(x))
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    files: getMultiFileValidation(config),
  });

interface ClaimLevelUploadSchemaExtraProps {
  config: IAppOptions;
  project: Pick<ProjectDto, "impactManagementParticipation">;
}

type ClaimLevelUploadSchemaType = ReturnType<typeof getClaimLevelUpload>;
const documentsErrorMap = makeZodI18nMap({ keyPrefix: ["documents"] });
const getClaimLevelUpload = ({ config, project }: ClaimLevelUploadSchemaExtraProps) =>
  z.object({
    form: z.union([z.literal(FormTypes.ClaimLevelUpload), z.literal(FormTypes.ClaimReviewLevelUpload)]),
    projectId: projectIdValidation,
    partnerId: partnerIdValidation,
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x =>
          (project.impactManagementParticipation === ImpactManagementParticipation.Yes
            ? allowedImpactManagementClaimDocuments
            : allowedClaimDocuments
          ).includes(x),
        )
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    periodId: periodIdValidation,
    files: getMultiFileValidation(config),
  });

type ClaimDetailLevelUploadSchemaType = ReturnType<typeof getClaimDetailLevelUpload>;
const getClaimDetailLevelUpload = ({ config, project }: ClaimLevelUploadSchemaExtraProps) =>
  z.object({
    form: z.literal(FormTypes.ClaimDetailLevelUpload),
    projectId: projectIdValidation,
    partnerId: partnerIdValidation,
    costCategoryId: costCategoryIdValidation,
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x =>
          (project.impactManagementParticipation === ImpactManagementParticipation.Yes
            ? allowedImpactManagementClaimDocuments
            : allowedClaimDocuments
          ).includes(x),
        )
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    periodId: periodIdValidation,
    files: getMultiFileValidation(config),
  });

const getPcrLevelUpload = ({ config }: { config: IAppOptions }) =>
  z.object({
    form: z.literal(FormTypes.PcrLevelUpload),
    projectId: projectIdValidation,
    projectChangeRequestIdOrItemId: pcrItemIdValidation,
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x => allowedPcrLevelDocuments.includes(x))
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    files: getMultiFileValidation(config),
  });

type PcrLevelUploadSchemaType = ReturnType<typeof getPcrLevelUpload>;

const projectLevelDelete = z.object({
  form: z.literal(FormTypes.ProjectLevelDelete),
  projectId: projectIdValidation,
  documentId: z.string(),
});

const partnerLevelDelete = z.object({
  form: z.literal(FormTypes.PartnerLevelDelete),
  projectId: projectIdValidation,
  partnerId: partnerIdValidation,
  documentId: z.string(),
});

type ClaimLevelDeleteSchemaType = typeof claimLevelDelete;
const claimLevelDelete = z.object({
  form: z.union([z.literal(FormTypes.ClaimLevelDelete), z.literal(FormTypes.ClaimReviewLevelDelete)]),
  projectId: projectIdValidation,
  partnerId: partnerIdValidation,
  periodId: periodIdValidation,
  documentId: z.string(),
});

const claimDetailLevelDelete = z.object({
  form: z.literal(FormTypes.ClaimDetailLevelDelete),
  projectId: projectIdValidation,
  partnerId: partnerIdValidation,
  periodId: periodIdValidation,
  costCategoryId: costCategoryIdValidation,
  documentId: z.string(),
});

const pcrLevelDelete = z.object({
  form: z.literal(FormTypes.PcrLevelDelete),
  projectId: projectIdValidation,
  projectChangeRequestIdOrItemId: z.union([pcrItemIdValidation, pcrIdValidation]),
  documentId: z.string(),
});

const projectOrPartnerLevelDelete = z.discriminatedUnion("form", [partnerLevelDelete, projectLevelDelete]);

export {
  getPcrLevelUpload,
  getProjectLevelUpload,
  projectLevelDelete,
  partnerLevelDelete,
  getClaimLevelUpload,
  getClaimDetailLevelUpload,
  claimDetailLevelDelete,
  claimLevelDelete,
  pcrLevelDelete,
  projectOrPartnerLevelDelete,
  getBankStatementUpload,
  documentsErrorMap,
};
export type {
  ProjectLevelUploadSchemaType,
  ClaimLevelUploadSchemaType,
  UploadBankStatementSchemaType,
  PcrLevelUploadSchemaType,
  ClaimDetailLevelUploadSchemaType,
  ClaimLevelDeleteSchemaType,
};
