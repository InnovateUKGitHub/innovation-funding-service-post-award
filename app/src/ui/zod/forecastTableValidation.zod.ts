import { makeZodI18nMap } from "@shared/zodi18n";
import {
  ClaimStatusGroup,
  getClaimStatusGroup,
} from "@ui/components/atomicDesign/organisms/forecasts/ForecastTable/getForecastHeaderContent";
import {
  mapToForecastTableDto,
  MapToForecastTableProps,
} from "@ui/components/atomicDesign/organisms/forecasts/ForecastTable/NewForecastTable.logic";
import { z, ZodIssueCode } from "zod";
import { FormTypes } from "./FormTypes";
import {
  booleanValidation,
  partnerIdValidation,
  profileIdValidation,
  projectIdValidation,
} from "./helperValidators.zod";
import { getGenericCurrencyValidation } from "./currencyValidator.zod";

type ForecastTableSchemaType = ReturnType<typeof getForecastTableValidation>["schema"];
const getForecastTableValidation = (data: Omit<MapToForecastTableProps, "clientProfiles">) => {
  const errorMap = makeZodI18nMap({ keyPrefix: ["forecastTable"], context: data });

  const schema = z
    .object({
      projectId: projectIdValidation,
      partnerId: partnerIdValidation,
      form: z.union([
        z.literal(FormTypes.ClaimForecastSaveAndContinue),
        z.literal(FormTypes.ClaimForecastSaveAndQuit),
        z.literal(FormTypes.ProjectSetupForecast),
        z.literal(FormTypes.ForecastTileForecast),
      ]),
      profile: z.record(
        profileIdValidation,
        getGenericCurrencyValidation({
          label: "forms.forecastTable.profile.label",
          min: -1_000_000_000,
          required: true,
        }),
      ),
      submit: booleanValidation,
    })
    .superRefine(({ profile: clientProfiles, form, submit }, { addIssue, path }) => {
      const table = mapToForecastTableDto({ ...data, clientProfiles });
      const nonForecastClaims = data.claimTotalProjectPeriods.filter(
        x => getClaimStatusGroup(x.status) !== ClaimStatusGroup.FORECAST,
      );
      const finalClaim = nonForecastClaims.find(claim => claim.isFinalClaim);

      if (form === FormTypes.ProjectSetupForecast) {
        if (submit) {
          for (const costCategory of table.costCategories) {
            if (costCategory.differentThanAllocatedCosts && !costCategory.isCalculated) {
              addIssue({
                code: ZodIssueCode.custom,
                path: [...path, "costCategory", costCategory.costCategoryId],
                params: {
                  costCategoryName: costCategory.costCategoryName,
                  i18n: "errors.forecasts_different",
                },
              });
            }
          }
        }
      } else {
        // Don't check if the total is too big for project setup spend profile
        // (there is already a check for each cost cat)
        if (table.totalRow.total > table.totalRow.golCost) {
          addIssue({
            code: ZodIssueCode.too_big,
            maximum: table.totalRow.golCost,
            inclusive: false,
            type: "number",
            path: [...path, "total"],
          });
        }
      }

      if (clientProfiles && finalClaim) {
        addIssue({
          code: ZodIssueCode.custom,
          params: {
            i18n: finalClaim.isApproved ? "errors.final_claim_already_approved" : "errors.final_claim",
          },
        });
      }
    });

  return { errorMap, schema };
};

export { getForecastTableValidation };
export type { ForecastTableSchemaType };
