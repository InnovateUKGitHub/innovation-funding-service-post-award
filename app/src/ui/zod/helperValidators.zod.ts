import { SalesforcePrefixes } from "@framework/constants/salesforceConstants";
import { ClientFileWrapper } from "@client/clientFileWrapper";
import { z, ZodIssueCode } from "zod";
import { IsomorphicFileWrapper } from "@server/apis/isomorphicFileWrapper";
import { validDocumentFilenameCharacters } from "@ui/validation/validators/documentUploadValidator";
import { getFileExtension, getFileName } from "@framework/util/files";
import { IAppOptions } from "@framework/types/IAppOptions";
import { IFileWrapper } from "@framework/types/fileWapper";

const projectIdValidation = z
  .string()
  .startsWith(SalesforcePrefixes.Acc_Project__c)
  .transform(x => x as ProjectId);

const pcrIdValidation = z
  .string()
  .startsWith(SalesforcePrefixes.Acc_ProjectChangeRequest__c)
  .transform(x => x as PcrId);

const pcrItemIdValidation = z
  .string()
  .startsWith(SalesforcePrefixes.Acc_ProjectChangeRequest__c)
  .transform(x => x as PcrItemId);

const partnerIdValidation = z
  .string()
  .startsWith(SalesforcePrefixes.Acc_ProjectParticipant__c)
  .transform(x => x as PartnerId);

const emptyStringToUndefinedValidation = z
  .string()
  .refine(x => x === "")
  .transform(() => undefined)
  .optional();

const emptyStringToNullValidation = z
  .string()
  .refine(x => x === "")
  .transform(() => null)
  .nullable();

const periodIdValidation = z.coerce
  .number()
  .int()
  .gt(0)
  .lt(500) // Assumption that a project has fewer than 500 periods.
  .transform(x => x as PeriodId);

const getSingleFileValidation = (options: IAppOptions) => {
  const { imageTypes, pdfTypes, presentationTypes, spreadsheetTypes, textTypes } = options.permittedTypes;
  const permittedFileTypes = [...pdfTypes, ...textTypes, ...presentationTypes, ...spreadsheetTypes, ...imageTypes];

  return z.custom<IFileWrapper>().superRefine((file, ctx) => {
    const basename = getFileName(file.fileName);
    const extension = getFileExtension(file.fileName);

    if (basename.length === 0 || extension.length === 0) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_basename_too_small",
        },
      });
    }

    if (!(file.fileName.length <= options.maxFileBasenameLength)) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_basename_too_big",
          count: options.maxFileBasenameLength,
        },
      });
    }

    if (!validDocumentFilenameCharacters.test(file.fileName)) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_name_invalid_characters",
        },
      });
    }

    if (!permittedFileTypes.includes(extension)) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_extension_invalid_type",
        },
      });
    }

    if (!(file.size > 0)) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_size_too_small",
        },
      });
    }

    if (!(file.size <= options.maxFileSize)) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_size_too_large",
          size: options.maxFileSize,
        },
      });
    }
  });
};

const getMultiFileValidation = (options: IAppOptions) =>
  z.preprocess((x: unknown) => {
    // Map to ClientFileWrapper/ServerFileWrapper
    if (Array.isArray(x) && x.every(x => x instanceof IsomorphicFileWrapper)) return x;
    if ("FileList" in globalThis && x instanceof FileList) return [...x].map(x => new ClientFileWrapper(x));
    return null;
  }, z.array(getSingleFileValidation(options)).min(1).max(options.maxUploadFileCount));

export {
  projectIdValidation,
  pcrIdValidation,
  pcrItemIdValidation,
  partnerIdValidation,
  periodIdValidation,
  emptyStringToUndefinedValidation,
  emptyStringToNullValidation,
  getSingleFileValidation,
  getMultiFileValidation,
};
