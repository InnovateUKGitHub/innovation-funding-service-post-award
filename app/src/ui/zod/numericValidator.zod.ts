import { isNil } from "lodash";
import { z, ZodIssueCode } from "zod";

const defaultMax = 999_999_999_999;
const defaultMin = 0;

/**
 * ### getNumericValidation
 *
 *
 * this will match a set of validations and
 * also error messaging to ensure consistency across the app
 * and compliance with GDS and is for evaluating numeric values
 *
 */
export const getNumberValidation = ({
  max = defaultMax,
  min = defaultMin,
  required,
}: {
  max?: number;
  min?: number;
  required: boolean;
}) =>
  z
    .any()
    .nullable()
    .optional()
    .superRefine((val, ctx) => {
      const isEmpty = isNil(val);
      if (required && isEmpty) {
        return ctx.addIssue({
          code: ZodIssueCode.custom,
          params: {
            generic: true,
            i18n: "errors.generic.number.required",
          },
        });
      }

      if (!isEmpty) {
        // Check if the number is valid
        if (isNaN(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              generic: true,
              i18n: "errors.generic.number.invalid_type",
            },
          });
        }

        // Make sure our number isn't too big
        if (val > max) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              count: max,
              generic: true,
              i18n: "errors.generic.number.too_big",
            },
          });
        }

        if (val < min) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              count: min,
              generic: true,
              i18n: "errors.generic.number.too_small",
            },
          });
        }
      }
    });
