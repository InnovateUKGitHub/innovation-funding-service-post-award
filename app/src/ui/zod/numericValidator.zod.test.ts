import { initStubTestIntl } from "@shared/initStubTestIntl";
import { getNumberValidation } from "./numericValidator.zod";
import { makeZodI18nMap } from "@shared/zodi18n";
import { z } from "zod";
import copy from "@copy/default.en-GB.json";

describe("number currency validator", () => {
  beforeAll(async () => {
    await initStubTestIntl({
      forms: {
        errors: {
          generic: {
            number: copy.forms.errors.generic.number,
          },
        },
        beans: {
          means: {
            heinz: {
              label: "New Jeans",
            },
          },
        },
      },
    });
  });

  const errorMap = makeZodI18nMap({ keyPrefix: ["beans", "means"] });

  test.each([
    ["accept 0", 0, true],
    ["accept 1", 1, true],
    ["accept 2", 2, true],
    ["accept 3", 3, true],
    ["accept 5", 5, true],
    ["accept 1.01", 1.01, true],
    ["accept 500", 500, true],
    ["decline too small", -10, false],
    ["decline too big", 20000000, false],
    ["decline infinity", Infinity, false],
    ["decline null", null, false],
    ["decline undefined", undefined, false],
    ["decline NaN", NaN, false],
  ])("%s when required", (name, input, accept) => {
    const parse = z
      .object({
        heinz: getNumberValidation({
          max: 1_000_000,
          min: 0,
          required: true,
        }),
      })
      .safeParse({ heinz: input }, { errorMap });
    expect(parse.success).toBe(accept);
    expect(parse as unknown).toMatchSnapshot();
  });

  test.each([
    ["accept 0", 0, true],
    ["accept 1", 1, true],
    ["accept 2", 2, true],
    ["accept 3", 3, true],
    ["accept 5", 5, true],
    ["accept 1.01", 1.01, true],
    ["accept 500", 500, true],
    ["decline too small", -10, false],
    ["decline too big", 20000000, false],
    ["decline infinity", Infinity, false],
    ["accept null", null, true],
    ["accept undefined", undefined, true],
    ["decline NaN", NaN, false],
  ])("%s when not required", (name, input, accept) => {
    const parse = z
      .object({
        heinz: getNumberValidation({
          max: 1_000_000,
          min: 0,
          required: false,
        }),
      })
      .safeParse({ heinz: input }, { errorMap });
    expect(parse.success).toBe(accept);
    expect(parse as unknown).toMatchSnapshot();
  });
});
