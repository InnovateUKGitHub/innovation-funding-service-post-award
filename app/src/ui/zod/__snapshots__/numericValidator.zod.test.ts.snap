// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`number currency validator accept 0 when not required 1`] = `
{
  "data": {
    "heinz": 0,
  },
  "success": true,
}
`;

exports[`number currency validator accept 0 when required 1`] = `
{
  "data": {
    "heinz": 0,
  },
  "success": true,
}
`;

exports[`number currency validator accept 1 when not required 1`] = `
{
  "data": {
    "heinz": 1,
  },
  "success": true,
}
`;

exports[`number currency validator accept 1 when required 1`] = `
{
  "data": {
    "heinz": 1,
  },
  "success": true,
}
`;

exports[`number currency validator accept 1.01 when not required 1`] = `
{
  "data": {
    "heinz": 1.01,
  },
  "success": true,
}
`;

exports[`number currency validator accept 1.01 when required 1`] = `
{
  "data": {
    "heinz": 1.01,
  },
  "success": true,
}
`;

exports[`number currency validator accept 2 when not required 1`] = `
{
  "data": {
    "heinz": 2,
  },
  "success": true,
}
`;

exports[`number currency validator accept 2 when required 1`] = `
{
  "data": {
    "heinz": 2,
  },
  "success": true,
}
`;

exports[`number currency validator accept 3 when not required 1`] = `
{
  "data": {
    "heinz": 3,
  },
  "success": true,
}
`;

exports[`number currency validator accept 3 when required 1`] = `
{
  "data": {
    "heinz": 3,
  },
  "success": true,
}
`;

exports[`number currency validator accept 5 when not required 1`] = `
{
  "data": {
    "heinz": 5,
  },
  "success": true,
}
`;

exports[`number currency validator accept 5 when required 1`] = `
{
  "data": {
    "heinz": 5,
  },
  "success": true,
}
`;

exports[`number currency validator accept 500 when not required 1`] = `
{
  "data": {
    "heinz": 500,
  },
  "success": true,
}
`;

exports[`number currency validator accept 500 when required 1`] = `
{
  "data": {
    "heinz": 500,
  },
  "success": true,
}
`;

exports[`number currency validator accept null when not required 1`] = `
{
  "data": {
    "heinz": null,
  },
  "success": true,
}
`;

exports[`number currency validator accept undefined when not required 1`] = `
{
  "data": {
    "heinz": undefined,
  },
  "success": true,
}
`;

exports[`number currency validator decline NaN when not required 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "generic": true,
      "i18n": "errors.generic.number.invalid_type"
    },
    "path": [
      "heinz"
    ],
    "message": "New Jeans must be a number."
  }
]],
  "success": false,
}
`;

exports[`number currency validator decline NaN when required 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "generic": true,
      "i18n": "errors.generic.number.invalid_type"
    },
    "path": [
      "heinz"
    ],
    "message": "New Jeans must be a number."
  }
]],
  "success": false,
}
`;

exports[`number currency validator decline infinity when not required 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "count": 1000000,
      "generic": true,
      "i18n": "errors.generic.number.too_big"
    },
    "path": [
      "heinz"
    ],
    "message": "New Jeans must be 1000000 or less."
  }
]],
  "success": false,
}
`;

exports[`number currency validator decline infinity when required 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "count": 1000000,
      "generic": true,
      "i18n": "errors.generic.number.too_big"
    },
    "path": [
      "heinz"
    ],
    "message": "New Jeans must be 1000000 or less."
  }
]],
  "success": false,
}
`;

exports[`number currency validator decline null when required 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "generic": true,
      "i18n": "errors.generic.number.required"
    },
    "path": [
      "heinz"
    ],
    "message": "Enter New Jeans."
  }
]],
  "success": false,
}
`;

exports[`number currency validator decline too big when not required 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "count": 1000000,
      "generic": true,
      "i18n": "errors.generic.number.too_big"
    },
    "path": [
      "heinz"
    ],
    "message": "New Jeans must be 1000000 or less."
  }
]],
  "success": false,
}
`;

exports[`number currency validator decline too big when required 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "count": 1000000,
      "generic": true,
      "i18n": "errors.generic.number.too_big"
    },
    "path": [
      "heinz"
    ],
    "message": "New Jeans must be 1000000 or less."
  }
]],
  "success": false,
}
`;

exports[`number currency validator decline too small when not required 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "count": 0,
      "generic": true,
      "i18n": "errors.generic.number.too_small"
    },
    "path": [
      "heinz"
    ],
    "message": "New Jeans must be 0 or more."
  }
]],
  "success": false,
}
`;

exports[`number currency validator decline too small when required 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "count": 0,
      "generic": true,
      "i18n": "errors.generic.number.too_small"
    },
    "path": [
      "heinz"
    ],
    "message": "New Jeans must be 0 or more."
  }
]],
  "success": false,
}
`;

exports[`number currency validator decline undefined when required 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "generic": true,
      "i18n": "errors.generic.number.required"
    },
    "path": [
      "heinz"
    ],
    "message": "Enter New Jeans."
  }
]],
  "success": false,
}
`;
